packages/storage/src/persisted.ts

```diff
@@ -152,33 +173,73 @@ export function makePersisted<T, O extends Record<string, any> = {}>(
   const deserialize: (data: string) => T = options.deserialize || JSON.parse.bind(JSON);
-  const init = storage.getItem(name, storageOptions);
   const set =
     typeof signal[0] === "function"
       ? (data: string) => (signal[1] as any)(() => deserialize(data))
       : (data: string) => (signal[1] as any)(reconcile(deserialize(data)));
   let unchanged = true;
 
-  if (init instanceof Promise) init.then(data => unchanged && data && set(data));
-  else if (init) set(init);
+  function fetchFromStorage() {
+    const value = storage.getItem(name, storageOptions);
+    if (value)
+      if (typeof value === "object" && "then" in value) // Check if value is a promise
+        value.then(data => unchanged && data && set(data));
+      else set(value);
+  }
+
+  fetchFromStorage() // Initialize the underlying signal
+
+  let broadcastChannel: BroadcastChannel | undefined
+
+  if (options.sync)
+    onMount(() => {
+      broadcastChannel = new BroadcastChannel(getSyncBroadcastName(name))
+      broadcastChannel.onmessage = async (event: MessageEvent) => {
+        if (event.type !== "message" && event.isTrusted)
+          return
+        const message = event.data as SyncBroadcastMessage
+        if (message.__sync_broadcast_message && message.name === name)
+          fetchFromStorage() // Update the value, because it changed
+      }
+      onCleanup(() => {
+        const tmp = broadcastChannel;
+        broadcastChannel = undefined
+        tmp?.close()
+      })
+    })
+
+  function onChanged() {
+    broadcastChannel?.postMessage({__sync_broadcast_message: true, name: name} as SyncBroadcastMessage)
+    unchanged = false;
+  }
 
   return [
     signal[0],
     typeof signal[0] === "function"
       ? (value?: T | ((prev: T) => T)) => {
-          const output = (signal[1] as Setter<T>)(value as any);
+        const output = (signal[1] as Setter<T>)(value as any);
 
-          if (value) storage.setItem(name, serialize(output), storageOptions);
-          else storage.removeItem(name, storageOptions);
-          unchanged = false;
-          return output;
-        }
+        if (value) storage.setItem(name, serialize(output), storageOptions);
+        else storage.removeItem(name, storageOptions);
+        onChanged()
+        return output;
+      }
       : (...args: any[]) => {
-          (signal[1] as any)(...args);
-          const value = serialize(untrack(() => signal[0] as any));
-          // @ts-ignore
-          storage.setItem(name, value, storageOptions);
-          unchanged = false;
-        },
+        (signal[1] as any)(...args);
+        const value = serialize(untrack(() => signal[0] as any));
+        // @ts-ignore
+        storage.setItem(name, value, storageOptions);
+        onChanged()
+      },
   ] as typeof signal;
 }
+
+
+interface SyncBroadcastMessage {
+  /**
+   * This property makes it easier for developers to understand the origin of the message if they accidentally use the
```

Comment:
That's a good idea, but couldn't we instead have a Set of registered names and warn users on re-use?

Reply:
what do mean by that?

Reply:
Shouldn't the name be bound to the key in which they property is persisted? So we should avoid binding the same property to different keys as well as different keys to the same property.